Kernel/Devices/Loop/LoopDevice.cpp

```diff
@@ -0,0 +1,135 @@
+    // just doesn't make sense (could be non-seekable files or char devices)
+    if (!custody->inode().metadata().is_regular_file())
+        return Error::from_errno(ENOTSUP);
+
+    // NOTE: We always take the LoopDevice::all_instances spinlock first
+    // because on LoopDevice::remove the device could be removed and
+    // invoke remove of the InodeContext immediately afterwards.
+    return LoopDevice::all_instances().with([&](auto& all_instances_list) -> ErrorOr<NonnullRefPtr<LoopDevice>> {
+        auto device = TRY(s_all_inode_backing_contexts->with([custody](auto& contexts_list) -> ErrorOr<NonnullRefPtr<LoopDevice>> {
+            auto context = TRY(create_inode_backing_context(contexts_list, custody->inode()));
+            return *TRY(DeviceManagement::try_create_device<LoopDevice>(context, s_loop_device_id.fetch_add(1)));
+        }));
+        all_instances_list.append(*device);
+        return device;
+    });
+}
+
+void LoopDevice::start_request(AsyncBlockDeviceRequest& request)
+{
+    auto work_item_creation_result = g_io_work->try_queue([this, &request]() {
+        if (request.request_type() == AsyncBlockDeviceRequest::RequestType::Read) {
+            auto result = m_inode_context->inode().read_bytes(request.block_index() * request.block_size(), request.buffer_size(), request.buffer(), nullptr);
+            if (result.is_error())
+                request.complete(AsyncDeviceRequest::Failure);
+            else
+                request.complete(AsyncDeviceRequest::Success);
+            return;
+        } else if (request.request_type() == AsyncBlockDeviceRequest::RequestType::Write) {
+            auto result = m_inode_context->inode().write_bytes(request.block_index() * request.block_size(), request.buffer_size(), request.buffer(), nullptr);
+            if (result.is_error())
+                request.complete(AsyncDeviceRequest::Failure);
+            else
+                request.complete(AsyncDeviceRequest::Success);
+            return;
+        }
+        VERIFY_NOT_REACHED();
+    });
+    if (work_item_creation_result.is_error())
+        request.complete(AsyncDeviceRequest::OutOfMemory);
+}
+
+bool LoopDevice::can_read(OpenFileDescription const&, u64) const
+{
+    return true;
+}
+
+bool LoopDevice::can_write(OpenFileDescription const&, u64) const
+{
+    return true;
+}
+
+ErrorOr<size_t> LoopDevice::read(OpenFileDescription& description, u64 offset, UserOrKernelBuffer& buffer, size_t size)
+{
+    return m_inode_context->inode().read_bytes(offset, size, buffer, &description);
+}
+
+ErrorOr<size_t> LoopDevice::write(OpenFileDescription& description, u64 offset, UserOrKernelBuffer const& buffer, size_t size)
+{
+    return m_inode_context->inode().write_bytes(offset, size, buffer, &description);
+}
+
+// FIXME: Allow passing different block sizes to the constructor
+LoopDevice::LoopDevice(NonnullRefPtr<InodeContext> inode_context, unsigned index)
+    : BlockDevice(20, index, 512)
```

Comment:
Can we make this major number less magic?

Reply:
Also not sure if the Filesystems block size can/should be used as the block size here or not

Reply:
> Also not sure if the Filesystems block size can/should be used as the block size here or not

What do you mean by that? it's a plain `BlockDevice`, so from which filesystem should I take the value of a block size?
The idea in the future is to allow specifying the block size via the `ioctl` on `/dev/devctl`, but for now this is hardcoded.

> Can we make this major number less magic?

Hmm, we could, but then we have many other devices that use this pattern with no issue. Maybe it's time to add `Kernel/API/MajorNumbers.h`?